src/components/CandidateContest.tsx

```diff
@@ -125,101 +132,110 @@ class CandidateContest extends React.Component<Props, State> {
-    const targetIsSelected = vote && value === vote.id
-    this.selectCandidate(
-      targetIsSelected
-        ? undefined
-        : contest.candidates.find(candidate => candidate.id === value)
-    )
+    const newVote = vote.filter(c => c.id !== id)
+    this.props.updateVote(contest.id, newVote)
+  }
+
+  public handleUpdateSelection = (event: InputEvent) => {
+    const { vote } = this.props
+    const id = (event.target as HTMLInputElement).value
+    const candidate = this.findCandidateById(vote, id)
+    if (!!candidate) {
+      if (candidate.isWriteIn) {
+        this.setState({ candidatePendingRemoval: candidate })
+      } else {
+        this.removeCandidateFromVote(id)
+      }
+    } else {
+      this.addCandidateToVote(id)
+    }
   }
 
-  public handleChangeVoteAlert = (attemptedVoteCandidateName: string) => {
-    this.setState({ attemptedVoteCandidateName })
+  public handleChangeVoteAlert = (
+    attemptedOverVoteCandidate: OptionalCandidate
+  ) => {
+    this.setState({ attemptedOverVoteCandidate })
   }
 
   public closeAttemptedVoteAlert = () => {
-    this.setState({ attemptedVoteCandidateName: '' })
+    this.setState({ attemptedOverVoteCandidate: undefined })
   }
 
-  public handleSelectWriteInCandidate = () => {
-    const { vote } = this.props
-    if (!!vote && vote.name === this.state.writeInCandidateName) {
-      this.selectCandidate(undefined)
-    } else {
-      this.toggleWriteInCandidateModal(true)
-    }
+  public confirmRemovePendingWriteInCandidate = () => {
+    this.removeCandidateFromVote(this.state.candidatePendingRemoval!.id)
+    this.clearCandidateIdPendingRemoval()
   }
-  public handleWriteInCandidateDisabledClick = () => {
-    const { vote } = this.props
-    const { writeInCandidateName } = this.state
-    if (!!vote && vote.name !== writeInCandidateName) {
-      this.handleChangeVoteAlert(writeInCandidateName || 'a write-in candidate')
-    }
+
+  public clearCandidateIdPendingRemoval = () => {
+    this.setState({ candidatePendingRemoval: undefined })
+  }
+
+  public initWriteInCandidate = () => {
+    this.toggleWriteInCandidateModal(true)
   }
 
   public normalizeName = (name: string) =>
     name
       .trim()
       .replace(/\t+/g, ' ')
       .replace(/\s+/g, ' ')
-  public selectWriteInCandidate = () => {
-    const writeInCandidateName = this.normalizeName(
+
+  public addWriteInCandidate = () => {
+    const { contest, vote } = this.props
+    const normalizedCandidateName = this.normalizeName(
       this.state.writeInCandidateName
     )
-    this.setState({ writeInCandidateName })
-    this.selectCandidate({
-      id: 'writeInCandidate',
-      name: writeInCandidateName,
-    })
+    this.props.updateVote(contest.id, [
+      ...vote,
+      {
+        id: `write-in__${camelCase(normalizedCandidateName)}`,
+        isWriteIn: true,
+        name: normalizedCandidateName,
+      },
+    ])
+    this.setState({ writeInCandidateName: '' })
```

Comment:
I'm still struggling with this empty string as null situation, but I'll leave it be until I have a better suggestion.